tests/python/pants_test/source/test_source_root.py

```diff
@@ -91,6 +91,48 @@ def root(path, langs):
     self.assertEqual(root('src/go/src', ('go',)),
                       trie.find('src/go/src/foo/bar/baz.go'))
 
+  def test_source_root_trie_traverse(self):
+    def make_trie() -> SourceRootTrie:
+      return SourceRootTrie(SourceRootFactory({
+      'jvm': ('java', 'scala'),
+      'py': ('python',)
+    }))
+
+    trie = make_trie()
+    self.assertEqual(set(), trie.traverse())
+
+    trie.add_pattern('src/*')
+    trie.add_pattern('src/main/*')
+    self.assertEqual({
+      ('src/*', ('jvm','py')),
+      ('src/main/*', ('jvm', 'py'))
+    }, trie.traverse())
+
+    trie = make_trie()
+    trie.add_pattern('*')
```

Comment:
Just FYI: the pattern format roughly matches git's globs: https://git-scm.com/docs/gitignore#_pattern_format, so a single `*` matches exactly one path component, and a double star (`**`) matches zero or more path components.

Reply:
Sorry... this comment was intended to be relevant to #8045, which I assume will be consuming these patterns.

Reply:
Yes, we've talked about how that will work. Basically, fixed patterns will be matched as-is, and wildcard patterns will be prefixed by "**/" and have the single * replaced by every allowed language.